Mathlib/Tactic/FunProp.lean

```diff
@@ -127,37 +137,213 @@ displays
+    ```
+
+    - Constant Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_const (y : Y) : Continuous (fun (x : X) => y) := ..
+    ```
+
+    - Composition Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_comp (f : Y → Z) (g : X → Y) (hf : Continuous f) (hg : Continuous g) :
+      Continuous (fun (x : X) => f (g x)) := ..
+    ```
+
+    - Apply Theorem
+      It can be either non-dependent version
+    ```lean
+    @[fun_prop]
+    theorem continuous_apply (a : α) : Continuous (fun f : (α → X) => f a) := ..
+    ```
+      or dependent version
+    ```lean
+    @[fun_prop]
+    theorem continuous_apply (a : α) : Continuous (fun f : ((a' : α) → E a') => f a) := ..
+    ```
+
+    - Pi Theorem
+    ```lean
+    @[fun_prop]
+    theorem continuous_pi (f : X → α → Y) (hf : ∀ a, Continuous (f x a)) :
+       Continuous (fun x a => f x a) := ..
+    ```
+
+    Not all of these theorems have to be provided, but at least the identity and composition
+    theorems should be.
+
+    You should also add theorems for `Prod.mk`, `Prod.fst`, and `Prod.snd`. Technically speaking,
+    they fall under the *function theorems* category, but without them, `fun_prop` can't function
+    properly. We are mentioning them as they are used together with *lambda theorems* to break
+    complicated expressions into a composition of simpler functions.
+    ```lean
+    @[fun_prop]
+    theorem continuous_fst (f : X → Y × Z) (hf : Continuous f) :
+        Continuous (fun x => (f x).fst) := ...
+    @[fun_prop]
+    theorem continuous_snd (f : X → Y × Z) (hf : Continuous f) :
+        Continuous (fun x => (f x).snd) := ...
+    @[fun_prop]
+    theorem continuous_prod_mk (f : X → Y) (g : X → Z) (hf : Continuous f) (hg : Continuous g) :
+        Continuous (fun x => (f x, g x)) := ...
+    ```
+
+- Function Theorems:
+    When `fun_prop` breaks complicated expression apart using *lambda theorems* it then uses
+    *function theorems* to prove that each piece is, for example, continuous.
+
+    The function theorem for `Neg.neg` and `Continuous` can be stated as:
+    ```lean
+    @[fun_prop]
+    theorem continuous_neg : Continuous (fun x => Neg.neg x) := ...
+    ```
+    or as:
+    ```lean
+    @[fun_prop]
+    theorem continuous_neg (f : X → Y) (hf : Continuous f) : Continuous (fun x => Neg.neg f x) := ...
+    ```
+    The first form is called *uncurried form* and the second form is called *compositional form*.
+    You can provide either form; it is mainly a matter of convenience. You can check if the form of
+    a theorem has been correctly detected by turning on the option:
+    ```lean
+    set_option Meta.Tactic.fun_prop.attr true
+    ```
+    If you really care that the resulting proof term is as short as possible, it is a good idea to
+    provide both versions.
+
+    One exception to this rule is the theorem for `Prod.mk`, which has to be stated in compositional
+    form.
```

Comment:
Because the theorem for `Prod.mk` is used to decompose more complicated functions into simpler. 

For example,
When you want to prove something about the function `(fun x => sin (exp x) + cos (exp x))` it get decomposed as
```
(fun x => x.1 + x.2)
∘
(fun x => (sin (exp x), cos (exp x)))
```
You should have exact theorem for `(fun x => x.1 + x.2)`. 

To deal with `fun x => (sin (exp x), cos (exp x))` you apply theorem for `Prod.mk` to decompose it into two problems, one for `fun x => sin (exp x)` and one for `fun x => cos (exp x)`.

Reply:
I added 
> This because this theorem together with *lambda theorems* is used to break expression to smaller pieces and `fun_prop`  assumes it is written in compositional form.

Does that make it a bit more clear?